src/charm.py

```diff
@@ -53,92 +55,85 @@ def __init__(self, *args):
+            self.framework.observe(self.on.peer_relation_departed, self.on_relation_departed)
+            self.framework.observe(self.on.peer_relation_departed, self.remove_departed_nodes)
+            self.framework.observe(self.on.peer_relation_departed, self.update_status)
+            self.framework.observe(self.on.leader_elected, self.remove_departed_nodes)
+            self.framework.observe(self.on.leader_elected, self.update_status)
+            self.framework.observe(self.on.microk8s_provides_relation_joined, self.add_node)
             self.framework.observe(
-                self.on.microk8s_provides_relation_changed, self._record_hostnames
+                self.on.microk8s_provides_relation_changed, self.record_hostnames
             )
             self.framework.observe(
-                self.on.microk8s_provides_relation_departed, self._on_relation_departed
+                self.on.microk8s_provides_relation_departed, self.on_relation_departed
             )
+            self.framework.observe(
+                self.on.microk8s_provides_relation_departed, self.remove_departed_nodes
+            )
+            self.framework.observe(self.on.microk8s_provides_relation_departed, self.update_status)
 
-    def _record_hostnames(self, event: Union[RelationChangedEvent, RelationJoinedEvent]):
-        for unit in event.relation.units:
-            hostname = event.relation.data[unit].get("hostname")
-            if hostname is not None:
-                self._state.hostnames[unit.name] = hostname
-
-    def _on_relation_departed(self, event: RelationDepartedEvent):
-        remove_hostname = self._state.hostnames.pop(event.departing_unit.name, None)
-        if not self.unit.is_leader():
-            return
-
-        if remove_hostname:
-            remove_nodes = self._get_peer_data("remove_nodes", [])
-            remove_nodes.append(remove_hostname)
-            self._set_peer_data("remove_nodes", remove_nodes)
-
-        self._on_config_changed(None)
-
-    def _worker_open_ports(self, _: InstallEvent):
-        self.unit.open_port("tcp", 80)
-        self.unit.open_port("tcp", 443)
-
-    def _control_plane_open_ports(self, _: InstallEvent):
-        self.unit.open_port("tcp", 80)
-        self.unit.open_port("tcp", 443)
-        self.unit.open_port("tcp", 16443)
-
-    def _on_remove(self, _: RemoveEvent):
+    def on_remove(self, _: RemoveEvent):
         try:
             microk8s.uninstall()
         except subprocess.CalledProcessError:
             LOG.exception("failed to remove microk8s")
 
-    def _announce_hostname(self, event: Union[RelationJoinedEvent, RelationChangedEvent]):
-        self._state.hostname = socket.gethostname()
-        self._state.hostnames[self.unit.name] = self._state.hostname
-        event.relation.data[self.unit]["hostname"] = self._state.hostname
+    def on_upgrade(self, _: UpgradeCharmEvent):
+        microk8s.upgrade()
```

Comment:
Shouldn't there be some form of "upgrade orchestration" here? Do we upgrade upgrade all nodes at the same time? Is this future work?

Reply:
Out of scope currently, added a TODO item to pick this up in the future, since we have to take care of possible rollbacks in the case of errors